### PR TITLE
Fix new warnings

### LINKED ISSUE #41901: Annotate platform-specific implementation assemblies with [SupportedOSPlatform]
With the introduction of the Platform Compatibility Analyzer and the supporting attributes, we can now mark APIs or assemblies as platform-specific.  The analyzer will flag any platform-neutral code that calls into platform-specific code and produce warnings. PR #41760 enabled the analyzer in the runtime repo and it addressed the warnings introduced.

Some of the warnings surfaced in platform-specific implementation assemblies and we suppressed they by using `Debug.Assert` statements.  Ideally, our platform-specific implementation assemblies would be marked with `[SupportedOSPlatform]` attributes to indicate the specific platform for that assembly.  However, we need to be careful not to let those platform-specific attributes get propagated into the corresponding reference assemblies.

This topic came up here: https://github.com/dotnet/runtime/pull/41760#discussion_r482493253.  We previously had a rough draft attempt of applying the attributes automatically based on the TFM in #41209.  We weren't confident in the approach though, so we chose to proceed with #41760 as-is and file this issue for a follow-up investigation to see if there's a better way to accomplish this through assembly-level attributes.

/cc @terrajobst

### PR DESCRIPTION
**Pull request by buyaa-n**
After enabling platform compat analyzer for projects not having access to `System.Runtime.InteropServices.RuntimeInformation, System.Runtime.InteropServices.OSPlatform` types revealed more warnings in runtime repo, fixing them

By adding `[SupportedOSPlatform("windows")]` attribute into AssemblyInfo.cs for windows only builds dropped warnings in libraries from 359 to 8, the remaining 8 are windows only APIs call found in multitargeted builds which are fixed/asserted/suppressed with this PR

For CoreLib 62 warnings found and seems there is no windows only build for CoreLib to inject `[SupportedOSPlatform("windows")]` attribute, so all 62 warnings are just asserted/suppressed with this PR

Related to `Ensure that each of those projects can produce warnings/errors from the analyzer during a normal build` task of #42335
Could fix https://github.com/dotnet/runtime/issues/41901

### CODE REVIEW

File: eng/versioning.targets
Comment (buyaa-n) [inline eng/versioning.targets:26]: This is for adding `[SupportedOSPlatform("windows")]` attribute in `AssemblyInfo.cs` of windows builds so that the analyzer wouldn't want for windows only API call

File: eng/versioning.targets
Comment (buyaa-n) [inline eng/versioning.targets:38]: Most likely we don't want to add the `UnsupportedOSPlatform` attribute for test projects, let me know if am wrong

File: src/libraries/Microsoft.Diagnostics.Tracing.EventSource.Redist/src/Microsoft.Diagnostics.Tracing.EventSource.Redist.csproj
Comment (buyaa-n) [inline src/libraries/Microsoft.Diagnostics.Tracing.EventSource.Redist/src/Microsoft.Diagnostics.Tracing.EventSource.Redist.csproj:8]: Because we are adding the `[SupportedOSPlatform("windows")]` for all windows build I needed to add the attributes for projects not having it, please let me know if there is other way

**Comment by bartonjs** [comment]
Crypto part LGTM (the CspParameters parameter is only used by a method that throws PNSE on Linux and is already marked as [SupportedOSPlatform("windows")]).

File: eng/versioning.targets
Comment (safern) [reply eng/versioning.targets:38]: Could we also move this condition to the target itself?

Comment (ViktorHofer) [reply eng/versioning.targets:38]: Jep that condition makes more sense on the target.

Comment (buyaa-n) [reply eng/versioning.targets]: Yep, moved

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (marek-safar) [inline src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: Why is this needed if the file .Windows.cs ?

Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: As mentioned [in my other commment](https://github.com/dotnet/runtime/pull/43097#discussion_r500433469) my effort of adding [SupportedOSPlatform("windows")] attribute in AssemblyInfo.cs of windows only build is worked only for libraries build, not for corelib build, seems they built differently and couldn't fund out how corelib windows (or any platform) specific build setup. It seems to me there is no windows specific build or so, so all warnings found in corelib are suppressed like this or using other suppressing ways

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComEventsInfo.cs
Comment (eerhardt) [inline src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComEventsInfo.cs:26]: Since this is COM, would it be more appropriate to mark the whole class as `[SupportedOSPlatform("windows")]`?

File: src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs
Comment (eerhardt) [inline src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs:270]: Since this is COM - would it be more appropriate to mark this whole class as only supported on Windows?

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > It seems to me there is no windows specific build or so

cc @ericstj @safern @ViktorHofer please correct me if am wrong and point me in the right direction

File: src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComEventsInfo.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComEventsInfo.cs:26]: Yes if "COM" is windows only that is the most appropriate way

File: src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComActivator.cs:270]: Will annotate come related ones as windows only, thanks

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (safern) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: corelib at least from coreclr defines `TargetsWindows/Unix/Browser`. So you could use those properties. See here: https://github.com/dotnet/runtime/blob/master/src/coreclr/dir.common.props#L55-L63

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: >  It seems to me there is no windows specific build or so, so all warnings found in corelib are suppressed like this or using other suppressing ways

There are OS specific builds of CoreLib but it doesn't multi-target as we do in libraries. Meaning it uses `TargetOS` and other properties that are passed in via a root build.

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: Also note that there are two versions of CoreLib, one in the coreclr directory and one in Mono.

Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: Thanks @safern @ViktorHofer i will give it try

Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > Also note that there are two versions of CoreLib, one in the coreclr directory and one in Mono.

I have run mono build, didn't see any warning, but I am not sure if the Mono build is using the analyzer i have modified

Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > corelib at least from coreclr defines TargetsWindows/Unix/Browser. So you could use those properties. See here:

@safern  I am actually using the same property (` <TargetsWindows Condition="'$(TargetOS)' == 'Windows_NT'">true</TargetsWindows> `) in `versionioning.targets` but it is not working for corelib, seems corelib build is not using (importing) `versionioning.targets` in its build, the question might be what else i can use instead of `versionioning.targets` 

> There are OS specific builds of CoreLib but it doesn't multi-target as we do in libraries. 

@ViktorHofer what i am trying to do is `adding [SupportedOSPlatform("windows")] attribute in AssemblyInfo.cs of windows only build` but i don't see windows specific output with `AssemblyInfo.cs` for CoreLib, i see it only for platform less build result. That means even there were `OS specific builds of CoreLib` if it is not producing `OS specific AssemblyInfo.cs` then this i cannot suppress the warnings for windows only build

Comment (ViktorHofer) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > seems corelib build is not using (importing) versionioning.targets in its build

correct, `versioning.targets` is a libraries specific msbuild file that isn't imported in CoreLib.

Comment (marek-safar) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > didn't see any warning, but I am not sure if the Mono build is using the analyzer

It should and if not it needs fixing

File: eng/versioning.targets
Comment (jeffhandley) [reply eng/versioning.targets:26]: Where is `TargetsWindows` true when `IsWindowsSpecific` not? Could we instead change when `IsWindowsSpecific` is set to capture those scenarios? What would be the tradeoffs between the two approaches?

**Review by jeffhandley — APPROVED** [review]
Looks good to me, but I'd also like to see @safern and/or @ViktorHofer sign off before merging.

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: > > didn't see any warning, but I am not sure if the Mono build is using the analyzer i have **modified**, or if they use same analyzer settings as in the https://github.com/dotnet/runtime/blob/master/eng/Analyzers.props#L9
> 
> It should and if not it needs fixing

I meant i am not sure if mono uses the same analyzer settings as in [libraries](https://github.com/dotnet/runtime/blob/master/eng/Analyzers.props#L9), therefore i am not sure if my changes affected mono build

**Comment by buyaa-n** [comment]
> Looks good to me, but I'd also like to see @safern and/or @ViktorHofer sign off before merging.

Sure, in addition to their review i also waiting for the [analyzer changes](https://github.com/dotnet/roslyn-analyzers/pull/4281) released for dogfooding, so that i can update the analyzer version here and see all CI legs pass green

File: eng/versioning.targets
Comment (buyaa-n) [reply eng/versioning.targets:26]: > Where is TargetsWindows true when IsWindowsSpecific not? 

As you know we added `<IsWindowsSpecific>true</IsWindowsSpecific>` for assemblies only supported on Windows,  but `TargetsWindows` can be true for windows only and multi-targeted assemblies, so if the assembly is multitargeted and has `Windows_NT` target then `TargetsWindows` will be true but `<IsWindowsSpecific>true</IsWindowsSpecific>` will be false. 

> Could we instead change when IsWindowsSpecific is set to capture those scenarios? What would be the tradeoffs between the two approaches?

I don't think so, `IsWindowsSpecific` is for annotating assembly as supported windows everywhere refs/assemblies/assembIyInfo wherever needed, but `TargetsWindows` adding the attribute only for windows targeted builds. With that being said we might could use on only `TargetsWindows`, but i am not sure if this value will be set for windows only assemblies

Comment (buyaa-n) [reply eng/versioning.targets:26]: > With that being said we might could use on only `TargetsWindows`, but i am not sure if `TargetsWindows` will always be set for windows only assemblies

If we only use `TargetsWindows` then the Supported attribute is not being generated for ref assemblies even for windows only assemblies, keeping as is

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: FYI Mono corelib is consuming the same analyzer versions, no warnings on the build, i can see the warnings if intentionally add platform-specific attributes

File: src/coreclr/dir.common.props
Comment (eerhardt) [inline src/coreclr/dir.common.props:73]: (nit) can this change be reverted?

File: src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs
Comment (eerhardt) [inline src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs:472]: Should the whole `QuotaControl` type be marked as `SupportedOSPlatformAttribute("windows")`?

https://docs.microsoft.com/en-us/dotnet/api/system.directoryservices.protocols.quotacontrol says

> The QuotaControl class is used to pass the security identifier (SID) of a security principle to retrieve quota-related data.

Would someone ever use this class outside of Windows?

cc @joperezr

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/StandardOleMarshalObject.Windows.cs
Comment (eerhardt) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/StandardOleMarshalObject.Windows.cs:24]: Should the whole `StandardOleMarshalObject` class be marked as `SupportedOSPlatformAttribute("windows")`? Just like COM, I think `Ole` is Windows-specific technology.

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs
Comment (eerhardt) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:83]: This is the only member (besides ctor) of this class. Should the whole class be marked `SupportedOSPlatformAttribute("windows")`?

File: src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs
Comment (joperezr) [reply src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs:472]: It is true that today this control is Windows specific, but I'm not sure I'm aware of the implications of adding this attribute to the whole class so I'm not sure how to move forward. What does the actual attribute do? is it just for documentation purposes or does the compiler look at it?

Comment (jeffhandley) [reply src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs:472]: The compiler does look at it by way of the Platform Compat Analyzer, which is enabled by default in .NET 5 projects.  Annotating the class would give a warning to any reference of the type in a call site reachable on non-Windows platforms.

Comment (joperezr) [reply src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs:472]: I see, in that case I think it would make sense to put the attribute in the whole class, since when using this control you really have to specify a SID, so we do want to keep it windows specific for now.

File: eng/Analyzers.props
Comment (safern) [inline eng/Analyzers.props]: IS this version correct? I would be surprised if our version was 56, so seems like a typo 😄

Comment (safern) [reply eng/Analyzers.props]: It seems my notification was not updated and pointed me to this change which is outdated 😄

File: src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs
Comment (buyaa-n) [reply src/libraries/System.DirectoryServices.Protocols/ref/System.DirectoryServices.Protocols.cs:472]: Will move the attribute to the type, thanks!

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/StandardOleMarshalObject.Windows.cs
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/StandardOleMarshalObject.Windows.cs:24]: I am not sure, the file has Windows prefix `StandardOleMarshalObject.Windows.cs` and there is a unix version `StandardOleMarshalObject.Unix.cs`, though unix version has no any implementation https://github.com/dotnet/runtime/blob/01e82df18426d2924bc1b6c53cf6de1b7752eadb/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/StandardOleMarshalObject.Unix.cs

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs
Comment (buyaa-n) [reply src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:83]: Could be

File: eng/Analyzers.props
Comment (ViktorHofer) [inline eng/Analyzers.props:9]: While you are already at it, can you please move these versions into Versions.props?

Comment (buyaa-n) [reply eng/Analyzers.props]: Done

**Review by eerhardt — APPROVED** [review]
LGTM - as long as @ViktorHofer and @safern approve of the build changes.

**Review by safern — APPROVED** [review]
LGTM

File: eng/Versions.props
Comment (ViktorHofer) [inline eng/Versions.props:59]: nit: I don't think these package live in Arcade (as you placed them into the Arcade dependencies section).

File: eng/versioning.targets
Comment (ViktorHofer) [inline eng/versioning.targets:26]: Just for my purposes to better understand this. What's the difference between `IsWindowsSpecific` and `TargetsWindows`?

**Review by ViktorHofer — APPROVED** [review]

File: eng/versioning.targets
Comment (safern) [reply eng/versioning.targets:26]: IsWindowsSpecific was a hook added so that projects that are only supported on windows add that property and get the SupportedPlatform attribute into the assembly. I think with this change we might no longer need that property as I would expect those to have a Windows configuration.

Comment (buyaa-n) [reply eng/versioning.targets:26]: > I think with this change we might no longer need that property as I would expect those to have a Windows configuration.

That idea came up in the above comments, and I tried it, when using only `TargetsWindows` the Supported attribute is not being generated for ref assemblies even for windows only assemblies, so kept it as is

https://github.com/dotnet/runtime/pull/43097/files#r502720394

File: src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs
Comment (buyaa-n) [reply src/coreclr/src/System.Private.CoreLib/src/System/Threading/ClrThreadPoolBoundHandle.Windows.cs:19]: I have added the attribute in the coreclr windows build and the build result looks good, warnings are gone. But when run libs build it used windows version of the coreclr build which having assembly level "Supported" on windows attribute (`runtime\artifacts\bin\coreclr\Windows_NT.x64.Debug\IL\System.Private.CoreLib.dll`) which would cause much more warnings. So seems for coreclr better not to add the attribute for targeted build

File: eng/Versions.props
Comment (buyaa-n) [reply eng/Versions.props:59]: Moving to the `Libraries dependencies` section then doesn't see anything else better fit

File: eng/versioning.targets
Comment (safern) [reply eng/versioning.targets:26]: Ah, right. That is because the ref assemblies never cross compile per runtime, so `TargetsWindows` will not be defined for them.

**Comment by buyaa-n** [comment]
The one CI leg failure is unrelated, merging as i need it for [handling browser failures ](https://github.com/dotnet/runtime/pull/43363/checks?check_run_id=1266565753)

**Pull request merged**
